pkg/notify/notifier/discord/discord.go

```diff
@@ -0,0 +1,196 @@
+		if tmplName != "" {
+			n.receiver.TmplName = tmplName
+		} else {
+			n.receiver.TmplName = DefaultTextTemplate
+		}
+	}
+
+	var err error
+	n.tmpl, err = notifierCtl.GetReceiverTmpl(n.receiver.TmplText)
+	if err != nil {
+		_ = level.Error(n.logger).Log("msg", "DiscordNotifier: create receiver template error", "error", err.Error())
+		return nil, err
+	}
+	return n, nil
+}
+
+func (n *Notifier) Notify(ctx context.Context, data *template.Data) error {
+
+	mentionedUsers := n.receiver.MentionedUsers
+	for i := range mentionedUsers {
+		if mentionedUsers[i] == "everyone" {
+			mentionedUsers[i] = "@everyone"
+			continue
+		}
+		mentionedUsers[i] = fmt.Sprintf("<@%s>", mentionedUsers[i])
+	}
+
+	mentionedRoles := n.receiver.MentionedRoles
+	for i := range mentionedRoles {
+		mentionedRoles[i] = fmt.Sprintf("<@&%s>", mentionedRoles[i])
+	}
+
+	atUsers := strings.Join(mentionedUsers, "")
+	atRoles := strings.Join(mentionedRoles, "")
+	var length int
+	if n.receiver.Type == nil || *n.receiver.Type == constants.DiscordContent {
+		length = MaxContentLength - len(atUsers) - len(atRoles)
+	} else {
+		length = EmbedLimit - len(atUsers) - len(atRoles)
+	}
+	messages, _, err := n.tmpl.Split(data, length, n.receiver.TmplName, "", n.logger)
+	if err != nil {
+		_ = level.Error(n.logger).Log("msg", "DiscordNotifier: generate message error", "error", err.Error())
+		return err
+	}
+
+	group := async.NewGroup(ctx)
+	if n.receiver.Webhook != nil {
+		for index := range messages {
+			group.Add(func(stopCh chan interface{}) {
+				msg := fmt.Sprintf("%s\n%s%s", messages[index], atUsers, atRoles)
+				stopCh <- n.sendTo(ctx, msg)
+			})
+		}
+	}
+	return group.Wait()
+}
+
+func (n *Notifier) sendTo(ctx context.Context, content string) error {
+
+	message := &Message{}
+	if n.receiver.Type == nil || *n.receiver.Type == constants.DiscordContent {
+		message.Content = content
+	} else {
+		message.Embeds = []Embeds{
+			{
+				Description: content,
+			},
+		}
+	}
+
+	send := func() (bool, error) {
+		url, err := n.notifierCtl.GetCredential(n.receiver.Webhook)
+		if err != nil {
+			return false, err
+		}
+		var buf bytes.Buffer
+		err = utils.JsonEncode(&buf, message)
+		if err != nil {
+			return false, err
+		}
+		request, err := http.NewRequestWithContext(ctx, http.MethodPost, url, &buf)
+		if err != nil {
+			return false, err
+		}
+		request.Header.Set("Content-Type", "application/json; charset=utf-8")
+		client := http.Client{}
+		resp, err := client.Do(request)
+		if err != nil {
+			return true, err
+		}
+		code := resp.StatusCode
+		level.Debug(n.logger).Log("msg", "DiscordNotifier", "response code:", code)
+
+		if code != http.StatusNoContent {
+			return false, fmt.Errorf("DiscordNotifier: send message error, code: %d", code)
+		}
+		return false, nil
+	}
+
+	start := time.Now()
+	defer func() {
+		_ = level.Debug(n.logger).Log("msg", "DiscordNotifier: send message", "used", time.Since(start).String())
+	}()
+
+	retry := 0
+	for {
```

Comment:
Will it retry infinitely under extreme conditions?

Reply:
Should we limit the number of times？

Reply:
Yes

Reply:
120 is too large, I think 3 is more suitable.